src/jvmTest/kotlin/bdd/SpaceEngineersCucumberTest.kt

```diff
@@ -100,6 +130,55 @@ class SpaceEngineersCucumberTest {
         )
     }
 
+    private fun observeBlocks(): List<SeBlock> {
+        return environment.observe(ObservationArgs(ObservationMode.BLOCKS)).allBlocks
+    }
+
+    private fun blockToGrind(): SeBlock {
+        return observeBlocks().first { it.id == context.lastNewBlockId }
+    }
+
+    @When("Character grinds to {double} integrity.")
+    fun character_grinds_until_to_integrity(integrity: Double) {
+        var currentIntegrity = blockToGrind().integrity
+        environment.equip(context.grinderLocation!!)
+        sleep(500)
+        environment.interact(InteractionArgs(InteractionType.BEGIN_USE))
+        while (currentIntegrity > integrity) {
+            currentIntegrity = blockToGrind().integrity
+        }
+        environment.interact(InteractionArgs(InteractionType.END_USE))
+    }
+
+    @When("Character grinds to {double}% integrity.")
+    fun character_grinds_until_to_integrity_percentage(percentage: Double) {
+        val blockToGrind = blockToGrind()
+        var currentIntegrity = blockToGrind.integrity
+        val integrity = blockToGrind.maxIntegrity * percentage * 0.01
+        environment.equip(context.grinderLocation!!)
+        sleep(500)
+        environment.interact(InteractionArgs(InteractionType.BEGIN_USE))
+        while (currentIntegrity > integrity) {
+            currentIntegrity = blockToGrind().integrity
+        }
+        environment.interact(InteractionArgs(InteractionType.END_USE))
+    }
+
+    @When("Character torches block back up to max integrity.")
+    fun character_torches_block_back_up_to_max_integrity() {
+        val blockToGrind = blockToGrind()
+        var currentIntegrity = blockToGrind.integrity
+        environment.equip(context.torchLocation!!)
+        sleep(500)
+        environment.interact(InteractionArgs(InteractionType.BEGIN_USE))
+        while (currentIntegrity < blockToGrind.maxIntegrity) {
+            currentIntegrity = blockToGrind().integrity
+
+        }
+        environment.interact(InteractionArgs(InteractionType.END_USE))
+    }
```

Comment:
Maybe one subjective thing. We can discuss it later (perhaps on a call).

I like more empty space. Similarly to written text where there are paragraphs. I would put an empty line before and after the while clause. And probably the equip call. 

It's a pity there's no "paragraph" concept in code which could be displayed according to the user settings...

Reply:
When usually this comes up, I think real reason behind is, that method is too complicated/too long and needs to be decomposed into smaller pieces.

I completely agree, that this bdd code is getting too big. Think some of these functionalities should go into character controller or layer above it, which does more complicated tasks like this fixing something up until it reaches some integrity value.

I don't think we need paragraph concepts and it's for the best. You can extract "paragraph" into it's own method :-)
Java actually has something like that btw:
https://howtodoinjava.com/java/basics/what-is-block-statement-in-java/